river/stats/cov.py

```diff
@@ -47,57 +66,71 @@ def __init__(self, ddof=1):
         self.ddof = ddof
         self.mean_x = stats.Mean()
         self.mean_y = stats.Mean()
-        self._C = 0
         self.cov = 0
 
     def update(self, x, y, w=1.0):
         dx = x - self.mean_x.get()
         self.mean_x.update(x, w)
         self.mean_y.update(y, w)
-        self._C += w * dx * (y - self.mean_y.get())
-        self.cov = self._C / max(1, self.mean_x.n - self.ddof)
+        self.cov += (
+            w * (dx * (y - self.mean_y.get()) - self.cov) / max(self.mean_x.n - self.ddof, 1)
+        )
+        return self
+
+    def revert(self, x, y, w=1.0):
+        dx = x - self.mean_x.get()
+        self.mean_x.revert(x, w)
+        self.mean_y.revert(y, w)
+        self.cov -= (
+            w * (dx * (y - self.mean_y.get()) - self.cov) / max(self.mean_x.n - self.ddof, 1)
+        )
         return self
 
     def update_many(self, X: np.ndarray, Y: np.ndarray):
         dx = X - self.mean_x.get()
         self.mean_x.update_many(X)
         self.mean_y.update_many(Y)
-        self._C += (dx * (Y - self.mean_y.get())).sum()
-        self.cov = self._C / max(1, self.mean_x.n - self.ddof)
+        self.cov += (dx * (Y - self.mean_y.get()) - self.cov).sum() / max(
+            self.mean_x.n - self.ddof, 1
+        )
         return self
 
     def get(self):
         return self.cov
 
-    def __iadd__(self, other):
+    def _iadd(self, other_mean_x, other_mean_y, other_n, other_cov, other_ddof):
         old_mean_x = self.mean_x.get()
         old_mean_y = self.mean_y.get()
         old_n = self.mean_x.n
 
         # Update mean estimates
-        self.mean_x += other.mean_x
-        self.mean_y += other.mean_y
+        self.mean_x._iadd(other_n, other_mean_x)
+        self.mean_y._iadd(other_n, other_mean_y)
 
         if self.mean_x.n <= self.ddof:
             return self
 
         # Scale factors
         scale_a = old_n - self.ddof
-        scale_b = other.mean_x.n - other.ddof
+        scale_b = other_n - other_ddof
 
         # Scale the covariances
-        self.cov = scale_a * self.cov + scale_b * other.cov
+        self.cov = scale_a * self.cov + scale_b * other_cov
         # Apply correction factor
         self.cov += (
-            (old_mean_x - other.mean_x.get())
-            * (old_mean_y - other.mean_y.get())
-            * ((old_n * other.mean_x.n) / self.mean_x.n)
+            (old_mean_x - other_mean_x)
+            * (old_mean_y - other_mean_y)
+            * ((old_n * other_n) / self.mean_x.n)
         )
         # Reapply scale
         self.cov /= self.mean_x.n - self.ddof
```

Comment:
Is there a corner case where that `max(self.mean_x.n - self.ddof, 1)` would be needed here too?

Reply:
Better be safe. Good catch!

Reply:
XD